subxt/src/backend/legacy/mod.rs

```diff
@@ -84,183 +85,249 @@ impl<T: Config + Send + Sync + 'static> Backend<T> for LegacyBackend<T> {
-            self.methods.chain_subscribe_finalized_heads().await?;
-
-        // Get the last finalized block immediately so that the stream will emit every finalized block after this.
-        let last_finalized_block_ref = self.latest_finalized_block_ref().await?;
-        let last_finalized_block_num = self
-            .block_header(last_finalized_block_ref.hash())
-            .await?
-            .map(|h| h.number().into());
-
-        // Fill in any missing blocks, because the backend may not emit every finalized block; just the latest ones which
-        // are finalized each time.
-        let sub = subscribe_to_block_headers_filling_in_gaps(
-            self.methods.clone(),
-            sub,
-            last_finalized_block_num,
-        );
-        let sub = sub.map(|r| {
-            r.map(|h| {
-                let hash = h.hash();
-                (h, BlockRef::from_hash(hash))
+        let this = LegacyBackend {
+            methods: self.methods.clone(),
+            storage_page_size: self.storage_page_size,
+        };
+
+        let retry_sub = retry_stream(move || {
+            let this = LegacyBackend {
+                methods: this.methods.clone(),
+                storage_page_size: this.storage_page_size,
+            };
+            Box::pin(async move {
+                let sub = this.methods.chain_subscribe_finalized_heads().await?;
+
+                // Get the last finalized block immediately so that the stream will emit every finalized block after this.
+                let last_finalized_block_ref = this.latest_finalized_block_ref().await?;
+                let last_finalized_block_num = this
+                    .block_header(last_finalized_block_ref.hash())
+                    .await?
+                    .map(|h| h.number().into());
+
+                // Fill in any missing blocks, because the backend may not emit every finalized block; just the latest ones which
+                // are finalized each time.
+                let sub = subscribe_to_block_headers_filling_in_gaps(
+                    this.methods.clone(),
+                    sub,
+                    last_finalized_block_num,
+                );
+                let sub = sub.map(|r| {
+                    r.map(|h| {
+                        let hash = h.hash();
+                        (h, BlockRef::from_hash(hash))
+                    })
+                });
+
+                Ok(StreamOf(Box::pin(sub)))
             })
-        });
-        Ok(StreamOf(Box::pin(sub)))
+        })
+        .await?;
+
+        Ok(retry_sub)
     }
 
     async fn submit_transaction(
         &self,
         extrinsic: &[u8],
     ) -> Result<StreamOfResults<TransactionStatus<T::Hash>>, Error> {
-        let sub = self
-            .methods
-            .author_submit_and_watch_extrinsic(extrinsic)
-            .await?;
+        let sub = retry(|| self.methods.author_submit_and_watch_extrinsic(extrinsic)).await?;
```

Comment:
Ok, so here we retry the initial future that submits the transaciton, but then don't retry the stream.

It might be safer to start by not retrying anything at all here to avoid this case:

1. User submits transaction
2. Disconnection happens while we are waiting for RPC response (but server got the TX)
3. On reconnect, we try submitting again and get an error like `TransactionStatus::Invalid`. User assumes that transaction failed, but actually the original TX did in fact get submitted and included.

So let's just return a disconnected error or whatever straight away, because then the user knows that their TX may or may not have made it to the server and can act accordingly?

Reply:
good catch, done